### ISSUE TITLE
CV AUC - Python vs R

### ISSUE THREAD

**Issue by neich10** [Microsoft/LightGBM#724]
Cv in R gives higher AUC than in python? (python 2.7)

In Python:
import pandas as pd
import lightgbm as lgb

df_train = pd.read_csv('../train.txt')   # size (50000,718)
df_test = pd.read_csv('../test.txt')     # size (50000,718)     

y_train = df_train['TARGET'].values
X_train = df_train.drop('TARGET', axis=1).values
y_test  = df_test['TARGET'].values
X_test  = df_test.drop('TARGET', axis=1).values

feature_names = list(df_train.drop('TARGET', axis=1).columns)

lgb_train = lgb.Dataset(X_train, y_train)
lgb_eval = lgb.Dataset(X_test, y_test)

params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'num_leaves': 128,
    'learning_rate': 0.02,
    'colsample_bytree': 0.25
}

lgb_cv = lgb.cv(params,
                lgb_train,
                num_boost_round=500,
                nfold=3,
                verbose_eval=True
                )
Python Ouput:
[500]   cv_agg's auc: 0.793803 + 0.00144072

In R:

library(Matrix)
library(R6)
library(lightgbm)

train   <- read.csv("../train.txt")  # same as above
train.y <- train$TARGET
test   <- read.csv("../test.txt")    # same as above
test.y <- test$TARGET

sparse.train <- sparse.model.matrix(TARGET ~ ., data = train)
dtrain <- lgb.Dataset(sparse.train, label = train.y)

sparse.test <- sparse.model.matrix(TARGET ~ ., data = test)
dtest <- lgb.Dataset(sparse.test, label = test.y)

params <- list(objective = "binary",
               num_leaves = 128,
               learning_rate = 0.02,
               colsample_bytree = 0.25
)

lgb_cv <- lgb.cv(params,
                  dtrain,
                  500,
                  nfold = 3,
                  eval = "auc",
                  verbose = 1)

R Output: 
[500]:	valid's auc:0.818324+0.00262014.

Then I train (same parameters above) with df_train and predict in df_test in both Python and R, and evaluate auc with IBM SPSS Modeler. Python´s model gives 0.819 and R´s model gives 0.818.

Why Python CV agg´s auc differs in 0.02 aprox in compare to R CV? Both with same parameters and same datasets.

Thank you!

**Comment by Laurae2** [comment]
Your CV folds are different, make sure you are using premade folds first. R / Python RNGs are not identical.

**Comment by neich10** [comment]
Actually I found the problem was that default values of the stratified parameter is TRUE in R and False in Python. My database was highly inbalanced, so that´s why AUC was lower when the stratified parameter value was set to False (Python default), and higher when it was set to TRUE (R default).

Thank you!

**Comment by Laurae2** [comment]
Self: to fix in a future PR for Python, stratified CV by default

**Comment by vivekk-ezdi** [comment]
lightgbm.cv can also be used fore regression tasks. In that due to default True behaviour of stratify, an error is thrown in sklearn StratifiedKFold. See the [question here](https://stackoverflow.com/questions/49774825/python-lightgbm-cross-validation-how-to-use-lightgbm-cv-for-regression/49775393#49775393).

If we are sticking with `True`, then we should add a check if `objective` is `'regression'` is not. If not, appropriate error should be thrown, or maybe `stratify` should be changed to `False` with a warning.
